pilot/pkg/xds/workloadentry_test.go

```diff
@@ -0,0 +1,272 @@
+			Weight:         1,
+			ServiceAccount: "sa-a",
+		},
+	}
+	wgA = config.Config{
+		Meta: config.Meta{
+			GroupVersionKind: gvk.WorkloadGroup,
+			Namespace:        "a",
+			Name:             "wg-a",
+			Labels: map[string]string{
+				"grouplabel": "notonentry",
+			},
+		},
+		Spec:   tmplA,
+		Status: nil,
+	}
+)
+
+func TestNonAutoregisteredWorklaods(t *testing.T) {
+	store := memory.NewController(memory.Make(collections.All))
+	ig := NewInternalGen(&DiscoveryServer{instanceID: "pilot-1"})
+	ig.Store = store
+	createOrFail(t, store, wgA)
+
+	cases := map[string]*model.Proxy{
+		"missing group":      {IPAddresses: []string{"1.2.3.4"}, Metadata: &model.NodeMetadata{Namespace: wgA.Namespace}},
+		"missing ip":         {Metadata: &model.NodeMetadata{Namespace: wgA.Namespace, AutoRegisterGroup: wgA.Name}},
+		"missing namespace":  {IPAddresses: []string{"1.2.3.4"}, Metadata: &model.NodeMetadata{AutoRegisterGroup: wgA.Name}},
+		"non-existent group": {IPAddresses: []string{"1.2.3.4"}, Metadata: &model.NodeMetadata{Namespace: wgA.Namespace, AutoRegisterGroup: "dne"}},
+	}
+
+	for name, tc := range cases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			ig.RegisterWorkload(tc, &Connection{proxy: tc, Connect: time.Now()})
+			items, err := store.List(gvk.WorkloadEntry, model.NamespaceAll)
+			if err != nil {
+				t.Fatalf("failed listing WorkloadEntry: %v", err)
+			}
+			if len(items) != 0 {
+				t.Fatalf("expected 0 WorkloadEntry")
+			}
+		})
+	}
+
+}
+
+func TestAutoregistrationLifecycle(t *testing.T) {
+	if features.WorkloadEntryCleanupGracePeriod != 200*time.Millisecond {
+		features.WorkloadEntryCleanupGracePeriod = 200 * time.Millisecond
+	}
+	ig1, ig2, store := setup(t)
+	stopped1 := false
+	stop1, stop2 := make(chan struct{}), make(chan struct{})
+	defer func() {
+		// stop1 should be killed early, as part of test
+		if !stopped1 {
+			close(stop1)
+		}
+	}()
+	defer close(stop2)
+	go ig1.Run(stop1)
+	go ig2.Run(stop2)
+
+	p := fakeProxy("1.2.3.4", wgA, "nw1")
+
+	t.Run("initial registration", func(t *testing.T) {
+		// simply make sure the entry exists after connecting
+		ig1.RegisterWorkload(p, &Connection{proxy: p, Connect: time.Now()})
+		checkEntryOrFail(t, store, wgA, p, ig1.Server.instanceID)
+	})
+	t.Run("fast reconnect", func(t *testing.T) {
+		t.Run("same instance", func(t *testing.T) {
+			// disconnect, make sure entry is still there with disconnect meta
+			ig1.QueueUnregisterWorkload(p)
+			time.Sleep(features.WorkloadEntryCleanupGracePeriod / 2)
+			checkEntryOrFail(t, store, wgA, p, "")
+			// reconnect, ensure entry is there with the same instance id
+			ig1.RegisterWorkload(p, &Connection{proxy: p, Connect: time.Now()})
+			checkEntryOrFail(t, store, wgA, p, ig1.Server.instanceID)
+		})
+		t.Run("different instance", func(t *testing.T) {
+			// disconnect, make sure entry is still there with disconnect metadata
+			ig1.QueueUnregisterWorkload(p)
+			time.Sleep(features.WorkloadEntryCleanupGracePeriod / 2)
```

Comment:
Replace the `time.Sleep`s with `retry.UntilSuccessOrFail`

Reply:
why do we sleep anyways? isn't QueueUnregisterWorkload bypassing the wait? Can we insert an old workload entry then queue it?

Reply:
nevermind, read it wrong